### PR TITLE
Implemented basic Mutable torch tensorrt module pipeline

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

# Refit state transition graph
image

### CODE REVIEW

File: examples/dynamo/mutable_torchtrt_module_example.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    'mutable': True
+}
+
+model = models.resnet18(pretrained=False).eval().to("cuda")
+model2 = models.resnet18(pretrained=True).eval().to("cuda")
+mutable_module = torch_trt.compile(inputs=inputs, 
+                                   module=model, 
+                                   **kwargs)
+
+# Save the graph module as an exported program
```
Comment (narendasan) [inline examples/dynamo/mutable_torchtrt_module_example.py:52]: Demo the saving and reloading feature. Make it clear that the restriction on use_python_runtime is only for serialization

File: py/torch_tensorrt/_compile.py [hunk 2]
```diff
@@ -237,22 +239,27 @@ def compile(
         )
         return compiled_fx_module
     elif target_ir == _IRType.dynamo:
+        if kwargs["mutable"]:
```
Comment (narendasan) [inline py/torch_tensorrt/_compile.py:242]: Move this stuff under dynamo.compile

Comment (narendasan) [reply py/torch_tensorrt/_compile.py:242]: What are the acceptable input module types for MutableModule? e.g. nn.Module

Comment (narendasan) [inline py/torch_tensorrt/_compile.py:242]: Figure out what the best way to access this feature is, since it only seems to work for nn.Module and not a lower level frontend

Comment (narendasan) [reply py/torch_tensorrt/_compile.py:242]:
```py

# Sets a flag that says it needs to be compiled 
trt_mod = torch_tensorrt.MutableTorchTensorRTModule(mod, ...)

# compile on first run 
trt_mod.forward()

Comment (narendasan) [reply py/torch_tensorrt/_compile.py:242]: RE: Serialization, think about how we serialize the state (do we need to recompile).

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import torch
+
+from torch_tensorrt.dynamo._refit import refit_module_weights
+import torch_tensorrt as torch_trt
+import torch
+from torch_tensorrt.dynamo.utils import prepare_inputs
+from torch_tensorrt.dynamo._tracer import trace as dynamo_trace
+from torch_tensorrt.dynamo._compiler import compile as dynamo_compile
+
+class RefitFlag:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:13]:
```py
class LivenessFlag(Enum):
    UNKNOWN=auto()
    NEEDS_REFIT=auto()
    NEEDS_RECOMPILE=auto()
    LIVE=auto()
 
    _state = UNKNOWN

    def set_state():
       ...

```

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        self.original_inputs = sample_inputs
+        if not isinstance(sample_inputs, collections.abc.Sequence):
+            sample_inputs = [sample_inputs]
+        self.sample_inputs = tuple(sample_inputs)
+        self.torchtrt_inputs = prepare_inputs(self.sample_inputs)
+        self.kwargs = kwargs
+        self.original_model = pytorch_model
+        self.pytorch_model = _make_refit_change_trigger(pytorch_model, self.refit_flag)
+        self.gm = None
+        self.enabled_precisions_set = enabled_precisions_set
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:37]: Have a compile CompilationSettings obj.

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            "dla_global_dram_size": dla_global_dram_size,
+            "dryrun": dryrun,
+            "hardware_compatible": hardware_compatible,
+            "timing_cache_path": timing_cache_path,
+        }
+
+        self.settings = CompilationSettings(**compilation_options)
+
+    def load_state_dict(self, state_dict: dict[str, Any]) -> None:
+        RefitState.set_state(RefitFlag.NEEDS_REFIT)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:131]: What happens if there is more than one mutable module?

File: py/torch_tensorrt/_compile.py (hunk 2 above)
Comment (cehongwang) [reply py/torch_tensorrt/_compile.py:242]: No recompilation is needed. Saving implemented

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py (hunk 5 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:133]: You are right!

Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py]: Fixed

**Review by narendasan — APPROVED** [review]
LGTM

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+        version_compatible: bool = _defaults.VERSION_COMPATIBLE,
+        optimization_level: Optional[int] = _defaults.OPTIMIZATION_LEVEL,
+        use_python_runtime: bool = _defaults.USE_PYTHON_RUNTIME,
+        use_fast_partitioner: bool = _defaults.USE_FAST_PARTITIONER,
+        enable_experimental_decompositions: bool = _defaults.ENABLE_EXPERIMENTAL_DECOMPOSITIONS,
+        dryrun: bool = _defaults.DRYRUN,
+        hardware_compatible: bool = _defaults.HARDWARE_COMPATIBLE,
+        timing_cache_path: str = _defaults.TIMING_CACHE_PATH,
+        **kwargs: Any,
+    ) -> None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:79]: Add a docstr

**Review by narendasan — COMMENTED** [review]
Tests: 
1. Basic E2E workflow (ingest model and run)
2. Serialization for both rts (C++ and python) - build, save and load
3. Test refitting behavior (changing weights via attr) 
4. Test recompile ???

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+def test_save():
+
+    torch.manual_seed(0)
+    inputs = [torch.rand((1, 3, 224, 224)).to("cuda")]
+
+    # %%
+    # Compile the module for the first time and save it.
+    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    compile_spec = {
+        "use_python": False,
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:62]: Skip if runtime is not available

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+
+        cls = self.__class__
+        self.__class__ = type(
+            self.original_model.__class__.__name__,
+            (cls, pytorch_model.__class__),
+            {},
+        )
+        self.init_finished = True
+
+    def store_state_dict_meta_data(self) -> None:
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:201]: metadata instead of meta_data
